p5/core/vertex.py

```diff
@@ -309,8 +355,7 @@ def end_shape(mode=""):
     :type mode: str
 
     """
-    global is_bezier, is_curve, is_quadratic, is_contour
-    assert not is_contour, "begin_contour called without calling end_contour"
```

Comment:
What was the reasoning behind removing this assert?

Reply:
This assertion seems a bit wrong. The default value of `is_contour` is `False` so this assertion would raise an error even when no `contour` calls are made. 

Also, `is_contour` whether the current shape has a contour in it. So, we would have to use a different variable to indicate whether `end_contour` was called or not.

Reply:
I think the default value of `False` actually does not trigger this assert, since `not is_contour` would evaluate to `True`.
I think the original `is_contour` had the function of denoting whether `begin_contour` has been called without `end_contour` being called. Without calling `end_contour`, it doesn't make sense to call `end_shape`, hence the original assert.
Under what condition did this assert trigger for you?

Reply:
`end_contour` resets the value to `False` after `begin_contour` in the old code. I have removed that in the current code, because we need to know whether `is_contour` was called or not to be passed onto `renderer.end_shape`

Reply:
Ah I see. I think the old `is_contour` is still useful as a variable that indicates whether we are in a contour, thus providing us a way to nudge users to not call `end_shape` without calling `end_contour` after every `begin_contour`. I think the use case you've mentioned sounds like that a new variable named `has_contour` might be more appropriate. What do you think?

Reply:
Yes, that sounds right. I have done that. Thanks!